library/x509_create.c

```diff
@@ -125,81 +125,288 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+{
+    int ret = 0;
+    /*
+     * `cur` points to the current character in the string being parsed.
+     * `d` acts as a write pointer into the temporary 'data' buffer which holds
+     * the unescaped and unquoted DN values found in 'name'.
+     */
+    int in_quot = 0;
+    size_t value_len = 0;
     char data[MBEDTLS_X509_MAX_DN_NAME_SIZE];
     char *d = data;
 
-    /* Clear existing chain if present */
-    mbedtls_asn1_free_named_data_list( head );
+    /*
+     * As defined by RFC 1779, the DN is constructed from
+     * <string> | <key> <optional-space> "=" <optional-space> <string>
+     * where optional space is
+     * ( <CR> ) *( " " )
+     *
+     * When parsing the value:
+     * 1. Skip whitespaces.
+     * 2. Check if it is a quoted string.
+     * 3. Read the characters, to evaluate as a valid value.
+     * 4. Skip whitespaces until reaching separator character (',').
+     */
+
+    /*
+     * Skip whitespaces
+     */
+    while( *cur < end && *( *cur ) == ' ' )
+    {
+        ( *cur )++;
+    }
 
-    while( c <= end )
+    if( *cur == end || *( *cur ) == ',' )
     {
-        if( in_tag && *c == '=' )
-        {
-            if( ( attr_descr = x509_attr_descr_from_name( s, c - s ) ) == NULL )
-            {
-                ret = MBEDTLS_ERR_X509_UNKNOWN_OID;
-                goto exit;
-            }
+        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+        goto exit;
+    }
+    /*
+     * Check if it is a quoted string
+     */
+    else if( *( *cur ) == '\"' )
+    {
+        in_quot = 1;
+        ( *cur )++;
+    }
 
-            oid = attr_descr->oid;
-            s = c + 1;
-            in_tag = 0;
-            d = data;
+    while( *cur < end )
+    {
+        /*
+         * Check whether reached end of value to evaluate,
+         * whether it is a closing quotation, or a separator character
+         */
+        if( in_quot == 1 && *( *cur ) == '\"' )
+        {
+            /*
+             * Increase the current position to after the closing quotation
+             */
+            ( *cur )++;
+            /*
+             * Mark that quotation has ended.
+             */
+            in_quot = 0;
+            break;
         }
-
-        if( !in_tag && *c == '\\' && c != end )
+        else if( in_quot == 0 && *( *cur ) == ',')
         {
-            c++;
+            break;
+        }
 
-            /* Check for valid escaped characters */
-            if( c == end || *c != ',' )
+        if( *cur != end && *( *cur ) == '\\' )
+        {
+            ( *cur )++;
+            ret = x509_dn_check_special_chars( cur, end );
+            if( ret != 0 )
             {
                 ret = MBEDTLS_ERR_X509_INVALID_NAME;
                 goto exit;
             }
         }
-        else if( !in_tag && ( *c == ',' || c == end ) )
+
+        /*
+         * Fill the current data pointer with the unescaped character.
+         */
+        *( d++ ) = *( *cur );
```

Comment:
Question: Do we need to check that we have not ran out of space in the buffer pointed to by `d` before writing it?

Reply:
I don't think so. Because we first fill `d` and then we increment it by one. After that, we check that if `d - data == MBEDTLS_X509_MAX_DN_NAME_SIZE` we return error.
So, there is already a check that it isn't increased.

But, perhaps there is a special corner case where there is a valid string, followed only by white spaces, exceeding MBEDTLS_X509_MAX_DN_NAME_SIZE. In this case, value_len will be correct, but we will still write to `d` the whitespace. 
So, a followup question is:
In this corner case, do we want to return an error, or do we just don't write anymore to d, and use the correct value indicated by `value_len`?
I favor the second option